File: Funcional/LogicaInvasores.hs
module LogicaInvasores (
    estadoInicialInvasores, moverInvasores, dispararInvasores, verificarColisoesTirosInvasores, atualizarInvasores
) where

import System.Random (randomRIO)
import Tipos
import Debug.Trace

estadoInicialInvasores :: EstadoInvasores
estadoInicialInvasores = EstadoInvasores {
    invasores = [Invasor x y 2 | x <- [-150, -75, 0, 75, 150], y <- [200, 150, 100]],
    direcaoInvasores = 1,
    tempoProximoTiro = 2,
    tirosInimigos = []
}


moverInvasores :: Float -> EstadoInvasores -> EstadoInvasores
moverInvasores dt estado@(EstadoInvasores invs dir tempo tiros) =
    let novosInvasores = map (\inv -> inv { posicaoInvasorX = posicaoInvasorX inv + velocidadeInvasor inv * dir * dt * 50 }) invs
        (novosInvasoresAjustados, novaDirecao) = ajustarMovimento dt novosInvasores dir
    in estado { invasores = novosInvasoresAjustados, direcaoInvasores = novaDirecao }

ajustarMovimento :: Float -> [Invasor] -> Float -> ([Invasor], Float)
ajustarMovimento dt invs dir
    | any (\inv -> posicaoInvasorX inv <= (-larguraJanela/2) || posicaoInvasorX inv >= (larguraJanela/2)) invs =
        let novosInvasores = map (\inv -> inv { posicaoInvasorY = posicaoInvasorY inv - 10 * dt }) invs  -- Desce 10 unidades por segundo
            -- Verifica se os invasores já estão muito baixo na tela
            invasoresAjustados = if any (\inv -> posicaoInvasorY inv <= (-alturaJanela/2 + 50)) novosInvasores
                                then invs  -- Mantém os invasores na posição atual (não desce mais)
                                else novosInvasores
        in (invasoresAjustados, -dir)  -- Inverte a direção horizontal
    | otherwise = (invs, dir)


dispararInvasores :: EstadoInvasores -> IO ([Tiro], EstadoInvasores)
dispararInvasores estado@(EstadoInvasores invs _ tempo tiros) = do
    if tempo <= 0
        then do
            invasorAleatorio <- escolherInvasorAleatorio invs
            let novoTiro = Tiro (posicaoInvasorX invasorAleatorio) (posicaoInvasorY invasorAleatorio) 5
            return ([novoTiro], estado { tempoProximoTiro = 2, tirosInimigos = novoTiro : tiros })
        else return ([], estado { tempoProximoTiro = tempo - 0.1 })

escolherInvasorAleatorio :: [Invasor] -> IO Invasor
escolherInvasorAleatorio invs = do
    indice <- randomRIO (0, length invs - 1)
    return (invs !! indice)

verificarColisoesTirosInvasores :: [Tiro] -> [Invasor] -> ([Tiro], [Invasor])
verificarColisoesTirosInvasores tiros invs =
    let tirosRestantes = filter (\tiro -> not (any (colideComTiroInvasor tiro) invs)) tiros
        invasoresRestantes = filter (\inv -> not (any (\tiro -> colideComTiroInvasor tiro inv) tiros)) invs
    in (tirosRestantes, invasoresRestantes)


colideComTiroInvasor :: Tiro -> Invasor -> Bool
colideComTiroInvasor (Tiro tx ty _) (Invasor ix iy _) =
    abs(tx - ix) < 20 && abs(ty - iy) < 20

atualizarInvasores :: Float -> EstadoInvasores -> IO (EstadoInvasores, [Tiro])
atualizarInvasores dt estado = do
    let estadoMovido = moverInvasores dt estado
    (tirosInimigos, estadoAtualizado) <- dispararInvasores estadoMovido
    return (estadoAtualizado, tirosInimigos)

File: Prolog/logicaInvasores.pl
:- module(logicaInvasores, [invasor/2, estado_inicial_invasores/1, atualizar_invasores/2]).

:- dynamic invasor/3.  % invasor(X, Y, Velocidade)
:- dynamic tiro/3.     % tiro(X, Y, Velocidade)
:- dynamic estado_invasores/4. 

% Constantes
largura_janela(400).
altura_janela(600).

invasor(Window, invasor(X, Y, _)) :-
    new(Invasor, box(40, 30)),
    send(Invasor, fill_pattern, colour(green)),
    send(Window, display, Invasor, point(X, Y)).

%Estado inicial
estado_inicial_invasores(Estado) :-
    findall(invasor(X, Y, 2), 
           (member(X, [-330, -255, -180, -105, 30, 105, 180, 255, 330]), 
            member(Y, [200, 150, 100])),
           Invasores),
    Estado = estado_invasores(Invasores, 1, 2, []).

atualizar_invasores(Estado, Estado).

%mover Invasores
mover_invasores(DT, Estado, NovoEstado) :-
    Estado = estado_invasores(Invs, Dir, Tempo, Tiros),
    maplist(ajustar_posicao(DT, Dir), Invs, NovosInvasores),
    ajustar_movimento(DT, NovosInvasores, Dir, InvasoresAjustados, NovaDir),
    NovoEstado = estado_invasores(InvasoresAjustados, NovaDir, Tempo, Tiros).

ajustar_posicao(DT, Dir, invasor(X, Y, V), invasor(NovoX, Y, V)) :-
    NovoX is X + V * Dir * DT * 50.

ajustar_movimento(DT, Invs, Dir, InvasoresAjustados, NovaDir) :-
    algum_invasor_fora_da_tela(Invs) ->
        maplist(descer_invasor(DT), Invs, InvsDescidos),
        (algum_invasor_muito_baixo(InvsDescidos) ->
            InvasoresAjustados = Invs,
            NovaDir is -Dir
        ;
            InvasoresAjustados = InvsDescidos,
            NovaDir is -Dir
        )
    ;
        InvasoresAjustados = Invs,
        NovaDir = Dir.

algum_invasor_fora_da_tela(Invs) :-
    largura_janela(Largura),
    MeiaLargura is Largura / 2,
    member(invasor(X, _, _), Invs),
    (X =< -MeiaLargura ; X >= MeiaLargura).

algum_invasor_muito_baixo(Invs) :-
    altura_janela(Altura),
    MeiaAltura is Altura / 2,
    member(invasor(_, Y, _), Invs),
    Y =< (-MeiaAltura + 50).

descer_invasor(DT, invasor(X, Y, V), invasor(X, NovoY, V)) :-
    NovoY is Y - 10 * DT.

%Disparar Invasores
disparar_invasores(Estado, Tiros, NovoEstado) :-
    Estado = estado_invasores(Invs, Dir, Tempo, TirosAtuais),
    (Tempo =< 0 ->
        escolher_invasor_aleatorio(Invs, InvasorAleatorio),
        InvasorAleatorio = invasor(X, Y, _),
        NovoTiro = tiro(X, Y, 5),
        Tiros = [NovoTiro],
        NovoEstado = estado_invasores(Invs, Dir, 2, [NovoTiro | TirosAtuais])
    ;
        Tiros = [],
        NovoTempo is Tempo - 0.1,
        NovoEstado = estado_invasores(Invs, Dir, NovoTempo, TirosAtuais)
    ).

escolher_invasor_aleatorio(Invs, InvasorAleatorio) :-
    length(Invs, N),
    N > 0,
    random(0, N, Indice),
    nth0(Indice, Invs, InvasorAleatorio).

%Verificar colisões
verificar_colisoes_tiros_invasores(Tiros, Invasores, TirosRestantes, InvasoresRestantes) :-
    filtrar_tiros_sem_colisao(Tiros, Invasores, TirosRestantes),
    filtrar_invasores_sem_colisao(Invasores, Tiros, InvasoresRestantes).

filtrar_tiros_sem_colisao([], _, []).
filtrar_tiros_sem_colisao([Tiro|Resto], Invasores, Filtrados) :-
    colide_com_invasor(Tiro, Invasores) ->
        filtrar_tiros_sem_colisao(Resto, Invasores, Filtrados)
    ;
        Filtrados = [Tiro|Outros],
        filtrar_tiros_sem_colisao(Resto, Invasores, Outros).

filtrar_invasores_sem_colisao([], _, []).
filtrar_invasores_sem_colisao([Inv|Resto], Tiros, Filtrados) :-
    colide_com_tiro(Inv, Tiros) ->
        filtrar_invasores_sem_colisao(Resto, Tiros, Filtrados)
    ;
        Filtrados = [Inv|Outros],
        filtrar_invasores_sem_colisao(Resto, Tiros, Outros).

colide_com_invasor(tiro(TX, TY, _), Invasores) :-
    member(invasor(IX, IY, _), Invasores),
    abs(TX - IX) < 20,
    abs(TY - IY) < 20.

colide_com_tiro(invasor(IX, IY, _), Tiros) :-
    member(tiro(TX, TY, _), Tiros),
    abs(TX - IX) < 20,
    abs(TY - IY) < 20.

%Atualizar Invasores
atualizar_invasores(DT, Estado, NovoEstado, TirosInimigos) :-
    mover_invasores(DT, Estado, EstadoMovido),
    disparar_invasores(EstadoMovido, TirosInimigos, NovoEstado).       